### PR TITLE
[Zoe] Refactor Zoe State and refine the use of promises

### LINKED ISSUE #100: [Zoe] Refactor state.js to be in the form of records
We should refactor state.js such that instead of a large number of WeakMaps, we have records for every abstraction that would be a separate table in a database.

### PR DESCRIPTION
**Pull request by katelynsills**
This PR completely refactors `state.js` to use a table-like format with record validation. It also eliminates `zoeUtils.js` and refactors Zoe to be cleaner and have better use of promises.

It does not change how `escrowReceipts` work, and does not change any of the Zoe contracts. The Zoe API for the user and the Zoe API to the smart contracts have all of the same methods as before, plus some forward-looking methods. For instance, methods were introduced to use units rather than extents, even though all contracts currently still use extents.

Closes #100

### CODE REVIEW

**Comment by katelynsills** [comment]
I'll be doing further clean up of the comments tomorrow morning, but wanted to get this up for review

File: packages/zoe/state.js
Comment (erights) [inline packages/zoe/state.js]: Consider putting the `handle` parameter last an have it default to a fresh empty pass-by-presence identity

```js
create: (record, handle = harden({})) => {
```

I expect that most uses would prefer to just have the table create the handle.

File: packages/zoe/state.js
Comment (erights) [inline packages/zoe/state.js:20]: I saw a recent discussion of style choices around things named `insist`, but I didn't follow it. Do we have a convention for when something like this is an insistSomething vs a validateSomething?

File: packages/zoe/state.js
Comment (erights) [inline packages/zoe/state.js:130]: These tests are not adequate to determine that you actually have a valid offer. Malformed offers will pass these tests. What do you intend here?

**Review by erights — COMMENTED** [review]
Getting comments so far out now because I have no more time before party. When I read through the rest of it, I might change my mind about earlier comments, but so it goes.

**Review by erights — CHANGES_REQUESTED** [review]

**Review by erights — CHANGES_REQUESTED** [review]
After a bit, I stopped making review comments inline, but rather prepared https://github.com/Agoric/agoric-sdk/pull/339 as an explanation of the kind of decolumnization I have in mind. More is needed.

I'll be in tomorrow with plenty of time to discuss.

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: Good idea

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: I looked, and in most cases, we only want the handle. There is one case in which we return more than the handle, but it will change with the invites PR.  I changed this to only return the handle with a comment explaining the unusual case and how it will change in the future.

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: That change is in

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js:20]: I modeled the table representation on what you might expect in an ORM interface. This step is usually called validation as far as I know: 

https://docs.djangoproject.com/en/3.0/ref/models/instances/#validating-objects
https://mongoosejs.com/docs/validation.html

The validation functions will likely always use `insist` functions internally. There didn't seem to be a reason to expose that name though.

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js:107]: Great, I added that as a TODO

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js:130]: Just recording what we talked over in person - I think we agreed that this PR isn't meant to protect against malformed offers, but that we should make a note of that. I changed the name of the `validate` functions to be `validateSomewhat` and added TODOs to cover protection against malformed offer

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: Sure, that's a good improvement. Just changed it

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js:95]: Added as a TODO

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js:5]: The npm package is named `@agoric/ertp` (lower case). [Upper case characters aren't allowed](https://github.com/npm/validate-npm-package-name#naming-rules).

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js:156]: Yup!

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js:103]: Yep, that's right.

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js]: Hmm, if we didn't flatten it, what would we call the property that it is under? Flattening seemed like the natural thing to do, to me. It seemed easier to look at. 

By the way, if we choose to change this, it will break the current invites as used by the covered call and we should do it as a separate PR.

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: Good point. I think this comment is old with the addition of `getOffers`. After merging in your PR to this one, there are no longer `getPayoutRules` or `getExitRule` or `getUnitMatrix`.

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js]: Good point. That comment has been added

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js]: Good call. Looks much cleaner now

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: Done in Mark's PR

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: Done in Mark's PR

File: packages/zoe/zoe.js
Comment (erights) [inline packages/zoe/zoe.js:206]: Since the offer safety of each offer is independent, you could decolumnize this further.

File: packages/zoe/zoe.js
Comment (erights) [inline packages/zoe/zoe.js:239]: Hard to read this and be certain that it is equivalent to a degenerate use of `escrowOffer`. Would prefer that it be coded as a wrapper around a call to `escrowOffer` so that would be clear.

Separately, we talked about getting rid of all methods on the facet that are redundant with methods on the zoeService, of which the escrowing methods stand out. Instead, we could add one `getZoeService()` method to the facet, so that a contract can generally act as a client of the service in any way it wants. However, we'll do that after this PR. (Worth adding a TODO or issue or something.)

When that happens, `escrowEmptyOffer()` would move to the service, but still as a trivial wrapper. Or perhaps a separate convenience function, keeping unnecessary code out of Zoe itself.

File: packages/zoe/zoe.js
Comment (erights) [inline packages/zoe/zoe.js]: Same issue as with `assayTable`. You're entering the offer into the table before it's ready. If you postpone until ready, you'll also be able to use correct values for the `units` and `extents`. Only then does the record uphold the intended invariants. Until then we should keep this out of shared tables.

File: packages/zoe/zoe.js
Comment (erights) [inline packages/zoe/zoe.js:298]: I think there are more decolumnization opportunities here that will make some of these go away. But such further progress could happen after this PR, since it is more entangled with other upcoming changes.

File: packages/zoe/zoe.js
Comment (erights) [inline packages/zoe/zoe.js:426]: Nice separation.

File: packages/zoe/zoe.js
Comment (erights) [inline packages/zoe/zoe.js]: Same issue above: prematurely entering non-ready offer into table.

Assuming most of this logic survives the upcoming receipt removal, please let's try to consolidate most of this escrowing logic somehow, even in this PR, because it's too hard to review for equivalence across changes.

OTOH, if the upcoming receipt removal really makes more of this code irrelevant, and if it is hard to consolidate, then don't.

**Review by erights — CHANGES_REQUESTED** [review]
Mostly minor. However, I do want to review the delaying logic for the offerTable before LGTMing.

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: Thanks for catching that I hadn't changed it for the `update` case. I've changed that now

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js]: sure, we can change it. I changed the name to `customProperties`

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: good catch

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js]: Yeah, I intended an empty object, not realizing that `...undefined` works. I'll change it to `() => undefined`

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js:228]: Thanks so much!

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js:50]: No problem! I curried it so that the sorting only happens once. I don't think it makes sense for the user to pre-sort the expected properties though - I could see someone easily forgetting and it causing errors. Would be possible to not harden this parameter, especially since it is only used by other state.js code?

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js:206]: good point. Let's revisit this when we rework the invariant checks when extents are turned into units.

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js:298]: Agreed

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js:426]: Thanks!

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js:239]: I factored out a method called `depositPayments` so that the duplicate code is de-duped. I like the `getZoeService` idea a lot.

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js]: I've changed the code so that we wait until the deposit succeeds before creating the offer record.

File: packages/zoe/zoe.js
Comment (erights) [inline packages/zoe/zoe.js]: Should `instanceHandle` still be undefined when it comes from the facet?

I noticed this only because you hadn't yet consolidated `escrowEmptyOffer`, so I was checking it for equivalence. It does initialize `instanceHandle`.

Also, just checking: When we get rid of receipts, then the `instanceHandle` will always be initialized before inserting the offerRecord into the table, right?

File: packages/zoe/zoe.js
Comment (erights) [inline packages/zoe/zoe.js:259]: What do you think about consolidating `escrowEmptyOffer` too? Is there a reason not to?

**Review by erights — APPROVED** [review]
LGTM modulo three suggestions

I like it!

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js:259]: Yeah, escrowEmptyOffer doesn't do any of the depositing of the payments, so there is little overlap apart from the creation of an offer record, and that is already one method call.

File: packages/zoe/state.js
Comment (katelynsills) [reply packages/zoe/state.js:50]: Ah, gotcha. I understand now. That works for me

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js]: Ah, good point. `instanceHandle` should be defined if the call is made from the smart contract. I changed `depositPayments` to take an optional `instanceHandle`

Comment (katelynsills) [reply packages/zoe/zoe.js]: And, yes! When we get rid of receipts, there is no escrow before choosing a contract instance to join.

File: packages/zoe/zoe.js
Comment (katelynsills) [reply packages/zoe/zoe.js:259]: On trying to use `depositPayment` in `escrowEmptyOffer` I remembered that `escrowEmptyOffer` is as it is now such that it is synchronous. If we switch to a `getZoeServices` approach, we should revisit this, possibly adding this method to the services

**Pull request merged**
